### ISSUE TITLE
How to save response content to file

### ISSUE THREAD

**Issue by Pureaznangel** [haga-rak/fluxzy.core#285]
I just don't know how I can capture and save the response of a request, for example, just to the abc.com host to a text file.

Would you give me an example code to do this? ﻿

Thanks,﻿

**Comment by Pureaznangel** [comment]
Hello Sir, currently I need only the response body, I would be great if I
could get both of them (headers and body).

Thanks,

**Comment by Pureaznangel** [comment]
Yes sir, I need to process the traffic it in realtime. How could I do this?
Can you give me an example? I currently using Fiddler Extension for this,
but they didnt support .Net 6,7,8 so I found your Library for my solutions,
I wish it could solve my problem.

Thanks,

**Comment by Pureaznangel** [comment]
No Sir, I just need to read it only.

Regards,

**Comment by Pureaznangel** [comment]
Thank you so much for this! I will give it a try in tonight when I back
home then let you know the result. I hope I could help you make this
library greater.

Regards,

**Comment by Pureaznangel** [comment]
> We'll think for a more straightforward way to do this. However, for now, you can use the following code snippet to access response body in real time.
> 
> ```cs
>  internal class Program
>  {
>      static async Task Main()
>      {
>          var fluxzySetting = FluxzySetting.CreateLocalRandomPort();
>          
>          // Avoid deciphering traffic that is not from abc.com
>          fluxzySetting.ConfigureRule()
>              .When(new HostFilter("abc.com", StringSelectorOperation.EndsWith)
>              {
>                  Inverted = true // negate the condition
>              })
>              .Do(new SkipSslTunnelingAction())
>              .When(new HostFilter("abc.com", StringSelectorOperation.EndsWith))
>              .Do(new ReadResponseBodyAction());
>          
>          // Create a proxy instance
>          await using (var proxy = new Proxy(fluxzySetting))
>          {
>              var endpoints = proxy.Run();
>              
>              await using var proxyRegistration = await SystemProxyRegistrationHelper.Create(endpoints.First());
> 
>              // Fluxzy is now registered as the system proxy, the proxy will revert
>              // back to the original settings when proxyRegistration is disposed.
> 
>              Console.WriteLine("Press any key to halt proxy and unregistered");
>              Console.ReadKey();
>          }
>      }
>  }
> 
>  internal class ReadResponseBodyAction: Fluxzy.Rules.Action
>  {
>      public override FilterScope ActionScope => FilterScope.ResponseHeaderReceivedFromRemote;
>      public override string DefaultDescription => nameof(ReadResponseBodyAction);
> 
>      public override ValueTask InternalAlter(
>          ExchangeContext context, Exchange? exchange, Connection? connection, FilterScope scope,
>          BreakPointManager breakPointManager)
>      {
>          context.RegisterResponseBodySubstitution(new ResponseBodySubstitution(exchange!));
> 
>          return new ValueTask();
>      }
>  }
> 
>  internal class ResponseBodySubstitution : IStreamSubstitution
>  {
>      private readonly Exchange _exchange;
> 
>      public ResponseBodySubstitution(Exchange exchange)
>      {
>          _exchange = exchange;
>      }
> 
>      public async ValueTask<Stream> Substitute(Stream originalStream)
>      {
>          // Copy original response to MemoryStream
>          var memoryStream = new MemoryStream();
>          await originalStream.CopyToAsync(memoryStream);
> 
> 
>          Console.WriteLine($"{_exchange.FullUrl}. Length: {memoryStream.Length}");
> 
>          memoryStream.Seek(0, SeekOrigin.Begin); 
>          return memoryStream;
>      }
>  }
> ```

It worked better than expected. While waiting for you to complete these ideas, I will temporarily use this code to handle the problem inside the core of my software. Hopefully, the next version will be better and have more available methods to use for many necessary cases. For now, I can delete the fiddlercore and use your library instead.

Thank you for your helping!

**Comment by Pureaznangel** [comment]
> We'll think for a more straightforward way to do this. However, for now, you can use the following code snippet to access response body in real time.
> 
> ```cs
>  internal class Program
>  {
>      static async Task Main()
>      {
>          var fluxzySetting = FluxzySetting.CreateLocalRandomPort();
>          
>          // Avoid deciphering traffic that is not from abc.com
>          fluxzySetting.ConfigureRule()
>              .When(new HostFilter("abc.com", StringSelectorOperation.EndsWith)
>              {
>                  Inverted = true // negate the condition
>              })
>              .Do(new SkipSslTunnelingAction())
>              .When(new HostFilter("abc.com", StringSelectorOperation.EndsWith))
>              .Do(new ReadResponseBodyAction());
>          
>          // Create a proxy instance
>          await using (var proxy = new Proxy(fluxzySetting))
>          {
>              var endpoints = proxy.Run();
>              
>              await using var proxyRegistration = await SystemProxyRegistrationHelper.Create(endpoints.First());
> 
>              // Fluxzy is now registered as the system proxy, the proxy will revert
>              // back to the original settings when proxyRegistration is disposed.
> 
>              Console.WriteLine("Press any key to halt proxy and unregistered");
>              Console.ReadKey();
>          }
>      }
>  }
> 
>  internal class ReadResponseBodyAction: Fluxzy.Rules.Action
>  {
>      public override FilterScope ActionScope => FilterScope.ResponseHeaderReceivedFromRemote;
>      public override string DefaultDescription => nameof(ReadResponseBodyAction);
> 
>      public override ValueTask InternalAlter(
>          ExchangeContext context, Exchange? exchange, Connection? connection, FilterScope scope,
>          BreakPointManager breakPointManager)
>      {
>          context.RegisterResponseBodySubstitution(new ResponseBodySubstitution(exchange!));
> 
>          return new ValueTask();
>      }
>  }
> 
>  internal class ResponseBodySubstitution : IStreamSubstitution
>  {
>      private readonly Exchange _exchange;
> 
>      public ResponseBodySubstitution(Exchange exchange)
>      {
>          _exchange = exchange;
>      }
> 
>      public async ValueTask<Stream> Substitute(Stream originalStream)
>      {
>          // Copy original response to MemoryStream
>          var memoryStream = new MemoryStream();
>          await originalStream.CopyToAsync(memoryStream);
> 
> 
>          Console.WriteLine($"{_exchange.FullUrl}. Length: {memoryStream.Length}");
> 
>          memoryStream.Seek(0, SeekOrigin.Begin); 
>          return memoryStream;
>      }
>  }
> ```

How could I get the response header and body sir, Please help.

**Comment by Pureaznangel** [comment]
> You can access the response headers with `exchange..GetResponseHeaders()`. This can be null if the remote close the connection before sending the response.
> 
> For example, if you want to access it in the ResponseBodySubstitution
> 
> ```cs
> internal class ResponseBodySubstitution : IStreamSubstitution
> {
>     private readonly Exchange _exchange;
> 
>     public ResponseBodySubstitution(Exchange exchange)
>     {
>         _exchange = exchange;
>     }
> 
>     public async ValueTask<Stream> Substitute(Stream originalStream)
>     {
>         // Copy original response to MemoryStream
>         var memoryStream = new MemoryStream();
>         await originalStream.CopyToAsync(memoryStream);
> 
>         // Response headers 
> 
>         var headers = _exchange.GetResponseHeaders()!;
> 
>         foreach (var header in headers)
>         {
>             Console.WriteLine($"{header.Name}: {header.Value}");
>         }
> 
>         Console.WriteLine($"{_exchange.FullUrl}. Length: {memoryStream.Length}");
> 
>         memoryStream.Seek(0, SeekOrigin.Begin);
>         return memoryStream;
>     }
> }
> ```
> 
> In this example also, the response body is in the `MemoryStream`. If you want a string out of it:
> 
> ```cs
> var responseByteArray = memoryStream.ToArray();
> var responseString = Encoding.UTF8.GetString(responseByteArray);
> ```

Success! Thank you very much. It's actually simpler than I thought.

**Comment by 0xStuart** [comment]
The body data I am seeing is unrecognisable.  I noticed there is a GetDecodedRequestBodyStream method, but it doesn't seem to work (or maybe I am using it wrong).

**Comment by 0xStuart** [comment]
Further to that, I have just noticed, sometimes the string 'responseString' text is legible, and sometimes it is not.  There doesn't seem to be anything obviously different, except randomness.  I'll do some more debugging...
All i am trying to do, is read the json I'm getting from a web site, and POST a message back, if it meets my criteria.
